### PR TITLE
Scripting: split out compile limits and caching

### PR DESCRIPTION
**Pull request by stu-elastic**
Phase 1 of adding compilation limits per context.
* Refactor rate limiting and caching into separate class, 
  `ScriptCache`,  which will be used per context.
* Disable compilation limit for certain tests.

Refs: #50152

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/script/ScriptService.java [hunk 1]
```diff
@@ -213,29 +200,23 @@ public ScriptService(Settings settings, Map<String, ScriptEngine> engines, Map<S
...
+        compiler = new ContextCompiler(
+                    SCRIPT_CACHE_SIZE_SETTING.get(settings),
+                    SCRIPT_CACHE_EXPIRE_SETTING.get(settings),
+                    compilationLimitsEnabled() ?
+                        SCRIPT_MAX_COMPILATIONS_RATE.get(settings):
+                        new Tuple<>(0, TimeValue.ZERO)
+                    );
+    }
+
+    boolean compilationLimitsEnabled() {
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/script/ScriptService.java]: Can you please add a comment explaining this is overridden for the test infrastructure?

File: server/src/test/java/org/elasticsearch/script/ScriptServiceTests.java [hunk 2]
```diff
@@ -102,25 +101,25 @@ StoredScriptSource getScriptFromClusterState(String id) {
...
     public void testCompilationCircuitBreaking() throws Exception {
         buildScriptService(Settings.EMPTY);
-        scriptService.setMaxCompilationRate(Tuple.tuple(1, TimeValue.timeValueMinutes(1)));
-        scriptService.checkCompilationLimit(); // should pass
-        expectThrows(CircuitBreakingException.class, () -> scriptService.checkCompilationLimit());
-        scriptService.setMaxCompilationRate(Tuple.tuple(2, TimeValue.timeValueMinutes(1)));
-        scriptService.checkCompilationLimit(); // should pass
-        scriptService.checkCompilationLimit(); // should pass
-        expectThrows(CircuitBreakingException.class, () -> scriptService.checkCompilationLimit());
+        scriptService.compiler.setMaxCompilationRate(Tuple.tuple(1, TimeValue.timeValueMinutes(1)));
```
Comment (rjernst) [inline server/src/test/java/org/elasticsearch/script/ScriptServiceTests.java]: This seems like a series of unit tests on ContextCompiler, not a test of ScriptService.  Can we add a ContextCompilerTests, and move this test there?

File: test/framework/src/main/java/org/elasticsearch/node/MockNode.java [hunk 3]
```diff
@@ -134,6 +138,14 @@ protected SearchService newSearchService(ClusterService clusterService, IndicesS
             bigArrays, fetchPhase, circuitBreakerService);
     }
 
+    @Override
+    protected ScriptService newScriptService(Settings settings, Map<String, ScriptEngine> engines, Map<String, ScriptContext<?>> contexts) {
+        if (getPluginsService().filterPlugins(MockScriptService.TestPlugin.class).isEmpty()) {
```
Comment (rjernst) [inline test/framework/src/main/java/org/elasticsearch/node/MockNode.java]: We need to actually add this test plugin to the mock plugins. See `ESIntegTestCase.getMockPlugins()` and ESSingleNodeTestCase.newNode(). For now we can probably just always add it.

**Review by rjernst — COMMENTED** [review]
Looks pretty good, just a few comments.

**Review by jdconrad — APPROVED** [review]
LGTM (once @rjernst 's comments have been addressed).

File: server/src/main/java/org/elasticsearch/script/ContextCompiler.java [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+import org.elasticsearch.common.cache.CacheBuilder;
+import org.elasticsearch.common.cache.RemovalListener;
+import org.elasticsearch.common.cache.RemovalNotification;
+import org.elasticsearch.common.collect.Tuple;
+import org.elasticsearch.common.unit.TimeValue;
+
+import java.util.Map;
+import java.util.Objects;
+
+public class ContextCompiler {
```
[the remark this answers is not in the document]
Comment (stu-elastic) [reply server/src/main/java/org/elasticsearch/script/ContextCompiler.java]: Added and renamed to `ScriptCache`.

File: server/src/main/java/org/elasticsearch/script/ScriptService.java (hunk 1 above)
Comment (stu-elastic) [reply server/src/main/java/org/elasticsearch/script/ScriptService.java]: Added.

File: server/src/test/java/org/elasticsearch/script/ScriptServiceTests.java (hunk 2 above)
Comment (stu-elastic) [reply server/src/test/java/org/elasticsearch/script/ScriptServiceTests.java]: Moved to `ScriptCacheTests`.

File: test/framework/src/main/java/org/elasticsearch/node/MockNode.java (hunk 3 above)
Comment (stu-elastic) [reply test/framework/src/main/java/org/elasticsearch/node/MockNode.java]: Added.

**Review by rjernst — APPROVED** [review]
LGTM

**Comment by stu-elastic** [comment]
Master: 0866031 
7.x: 376932a

**Pull request merged**